include/unifex/async_scope.hpp

```diff
@@ -0,0 +1,248 @@
+#include <unifex/inplace_stop_token.hpp>
+#include <unifex/just.hpp>
+#include <unifex/manual_lifetime.hpp>
+#include <unifex/receiver_concepts.hpp>
+#include <unifex/scheduler_concepts.hpp>
+#include <unifex/sender_concepts.hpp>
+#include <unifex/sequence.hpp>
+#include <unifex/transform.hpp>
+#include <unifex/type_traits.hpp>
+
+#include <atomic>
+#include <memory>
+
+#include <unifex/detail/prologue.hpp>
+
+namespace unifex {
+
+namespace _async_scope {
+
+struct async_scope;
+
+struct _receiver_base {
+  [[noreturn]] void set_error(std::exception_ptr) noexcept {
+    std::terminate();
+  }
+
+  friend inplace_stop_token
+  tag_invoke(tag_t<get_stop_token>, const _receiver_base& r) noexcept {
+    return r.stopToken_;
+  }
+
+  inplace_stop_token stopToken_;
+  void* op_;
+  async_scope* scope_;
+};
+
+template <typename Scheduler>
+struct _has_scheduler {
+  struct type;
+};
+
+template <typename Scheduler>
+using has_scheduler = typename _has_scheduler<Scheduler>::type;
+
+template <typename Scheduler>
+struct _has_scheduler<Scheduler>::type {
+  friend const Scheduler&
+  tag_invoke(tag_t<get_scheduler>, const type& r) noexcept {
+    return r.scheduler_;
+  }
+
+  UNIFEX_NO_UNIQUE_ADDRESS Scheduler scheduler_;
+};
+
+template <typename Sender, typename Scheduler>
+struct _receiver {
+  struct type;
+};
+
+template <typename Sender, typename Scheduler>
+using receiver = typename _receiver<Sender, Scheduler>::type;
+
+void record_done(async_scope*) noexcept;
+
+template <typename Sender, typename Scheduler>
+struct _receiver<Sender, Scheduler>::type final
+    : _receiver_base, has_scheduler<Scheduler> {
+  explicit type(
+      const Scheduler& scheduler,
+      inplace_stop_token stoken,
+      void* op,
+      async_scope* scope) noexcept
+      : _receiver_base{stoken, op, scope},
+        has_scheduler<Scheduler>{scheduler} {}
+
+  // receivers uniquely own themselves; we don't need any special move-
+  // construction behaviour, but we do need to ensure no copies are made
+  type(type&&) noexcept = default;
+
+  ~type() = default;
+
+  // it's just simpler to skip this
+  type& operator=(type&&) = delete;
+
+  void set_value() noexcept {
+    set_done();
+  }
+
+  void set_done() noexcept {
+    // we're about to delete this, so save the scope for later
+    auto scope = scope_;
+
+    using op_t = manual_lifetime<connect_result_t<Sender, type>>;
+
+    auto op = static_cast<op_t*>(op_);
+    op->destruct();
+    delete op;
+
+    record_done(scope);
+  }
+};
+
+struct async_scope {
+  async_scope() noexcept = default;
+
+  ~async_scope() {
+    [[maybe_unused]] auto state = opState_.load(std::memory_order_relaxed);
+
+    assert(is_stopping(state));
+    assert(op_count(state) == 0);
+  }
+
+  template (typename Sender, typename Scheduler)
+    (requires scheduler<Scheduler> AND
+     sender_to<Sender, receiver<Sender, remove_cvref_t<Scheduler>>>)
+  void submit(Sender&& sender, Scheduler&& scheduler) {
```

Comment:
This shouldn't be called `submit`. `submit` in unifex means something else. I think `spawn` would be a good name for this.